### PR TITLE
Fix mqtt retain related issues

### PR DESCRIPTION
**Pull request by pumudu88**
This PR will add cassandra CQL implementation for mqtt retain and fix minor issues in retain feature.

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java [hunk 1]
```diff
@@ -121,11 +121,38 @@
                     " WHERE " + CQLConstants.QUEUE_NAME + "=? AND " +
                     CQLConstants.MESSAGE_ID + "=?";
 
+
+    /**
+     * CQL prepared statement to insert retain metadata
```
Comment (indikasampath2000) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java]: prepared statement for update metadata right ?

**Comment by ramith** [comment]
Looks good! others please review..

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java [hunk 2]
```diff
@@ -882,49 +931,246 @@ public boolean isOperational(String testString, long testTime) {
...
+
+        execute(batchStatement, "update retain message content and metadata for topic " +
+                                metadata.getDestination());
+
+    }
+
+    /**
+     * Create new retain entry for given topic message
+     *
+     * @param message
```
Comment (hemikak) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java]: Need description for parameter.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java [hunk 3]
```diff
@@ -1015,4 +1261,32 @@ public void close() throws AndesException {
...
+        /**
+         * Retained message ID for the topic
+         */
+        public long messageID;
+
+        /**
+         * Retained item entry
+         *
+         * @param topicID
+         * @param messageID
```
Comment (hemikak) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java]: Need descriptions for parameter.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/inbound/InboundSubscriptionEvent.java [hunk 4]
```diff
@@ -143,19 +156,6 @@ private void handleOpenSubscriptionEvent() {
...
-                    + getSubscriptionID(), e);
-        } catch (SubscriptionAlreadyExistsException e) {
-            // exception will be handled by receiver
-            future.setException(e);
-        } finally {
-            future.set(isComplete);
-        }
     }
 
     public void prepareForNewSubscription(AndesSubscriptionManager subscriptionManager) {
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/inbound/InboundSubscriptionEvent.java]: Method comments missing.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java (hunk 2 above)
Comment (pumudu88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java]: Added in following PR.
https://github.com/wso2/andes/pull/154

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java (hunk 3 above)
Comment (pumudu88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java]: Added in following PR.
https://github.com/wso2/andes/pull/154

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/inbound/InboundSubscriptionEvent.java (hunk 4 above)
Comment (pumudu88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/inbound/InboundSubscriptionEvent.java]: Added in following PR.
https://github.com/wso2/andes/pull/154

**Pull request merged**